File: Python/Algorithm/Lowest_Highest_Prime.py
# Complexity: O(n*(log(log(n))))

lh = [(0, 0) for _ in range(N)]; sv = [False] * N
def lhp():
    lh[2] = (2, 2)
    for i in range(4, N, 2):
        sv[i] = True; lh[i] = (2, 2)
    for i in range(3, N, 2):
        if not sv[i]:
            lh[i] = (i, i)
            for j in range((i << 1), N, i):
                sv[j] = True
                if not lh[j][0]: lh[j] = (i, lh[j][1])
                lh[j] = (lh[j][0], i)

File: CPP/Algorithm/Lowest_Highest_Prime.cpp
// Complexity: O(n*(log(log(n))))

pair<int64_t, int64_t> lh[N]; bitset<N> sv;
void pre() {
    lh[2].first = lh[2].second = 2;
    for (int i = 4; i < N; i += 2) {
        sv[i] = true; lh[i].first = lh[i].second = 2;
    }
    for (int i = 3; i < N; i += 2) {
        if (!sv[i]) {
            lh[i].first = lh[i].second = i;
            for (int j = (i << 1); j < N; j += i) {
                sv[j] = true;
                if (!lh[j].first) lh[j].first = i;
                lh[j].second = i;
            }
        }
    }
}